### PR TITLE
Use predicate cache for all non-timestamp predicates

### PR DESCRIPTION
**Pull request by PSeitz-dd**
This is done to better support monitor queries which shift their timerange.

## Summary

- separate timestamp constraints from cacheable predicates during per-split request rewriting
- populate predicate-cache entries across the complete split and reuse them for partial and fully covered time windows
- avoid caching timestamp-only and other trivial `match_all`/`match_none` boolean predicates
- skip posting-list warmup on predicate-cache hits and bypass predicate-cache processing when disabled
- use deterministic term-set serialization and preserve required-term metadata through cache fillers
- report rejected cache admission and warn when a computed entry cannot be retained

## Implementation

Timestamp ranges are extracted from strict positive clauses and intersected with request and split bounds. The remaining non-timestamp predicate is cached, while any residual timestamp range is applied outside the cache node. Boolean predicates made trivial by timestamp extraction are simplified before cache-node insertion.

image

image

### CODE REVIEW

File: quickwit/quickwit-search/src/leaf_cache.rs [hunk 1]
```diff
@@ -247,8 +248,13 @@ impl PredicateCacheImpl {
                 &quickwit_storage::metrics::PREDICATE_CACHE,
             ),
             key_hasher: CacheKeyHasher::random(),
+            enabled: config.capacity().as_u64() > 0,
```
[the remark this answers is not in the document]
Comment (PSeitz) [reply quickwit/quickwit-search/src/leaf_cache.rs:251]: no, there should be a way to disable to functionality. Which is currently to set the predicate cache to 0.

File: quickwit/quickwit-search/src/leaf.rs [hunk 2]
```diff
@@ -915,6 +939,22 @@ async fn leaf_search_single_split(
...
+                    predicate_warmup_query.count(&searcher).inspect_err(|_| {
+                        leaf_search_state_guard
+                            .set_state(SplitSearchState::Error(SplitSearchErrorKind::TantivySearch))
+                    })?;
+                    drop(_predicate_warmup_span_guard);
+                    ctx_clone
+                        .doc_mapper
+                        .query(split_schema, query_ast.clone(), false, predicate_cache)
+                        .map_err(|error| TantivyError::InvalidArgument(error.to_string()))?
+                        .0
```
[the remark this answers is not in the document]
Comment (PSeitz) [reply quickwit/quickwit-search/src/leaf.rs:954]: added a warning log for that case

File: quickwit/quickwit-search/src/leaf.rs [hunk 3]
```diff
@@ -729,22 +729,46 @@ async fn leaf_search_single_split(
...
+                ctx.searcher_context.predicate_cache.clone() as Arc<dyn PredicateCache>,
+                split.split_id.clone(),
+            ))
+        };
+    let predicate_cache_miss_ast = predicate_cache
+        .as_ref()
+        .and_then(|(cache, cache_split_id)| {
+            let timestamp_field = ctx.doc_mapper.timestamp_field_name()?;
+            let predicate_ast = time_bounded_cached_predicate(&query_ast, timestamp_field)?;
+            let predicate_key = serde_json::to_string(&predicate_ast).ok()?;
```
[the remark this answers is not in the document]
Comment (PSeitz) [reply quickwit/quickwit-search/src/leaf.rs:746]: changed to `BTreeMap`

File: quickwit/quickwit-search/src/leaf.rs [hunk 4]
```diff
@@ -1272,6 +1274,27 @@ impl QueryAstTransformer for RemoveTimestampRange<'_> {
...
+            .chain(&bool_query.filter)
+            .any(|query_ast| matches!(query_ast, QueryAst::MatchNone))
+        {
+            return Ok(Some(QueryAst::MatchNone));
+        }
+        let only_matches_all = bool_query
+            .must
+            .iter()
+            .chain(&bool_query.filter)
+            .all(|query_ast| matches!(query_ast, QueryAst::MatchAll));
```
[the remark this answers is not in the document]
Comment (PSeitz) [reply quickwit/quickwit-search/src/leaf.rs:1289]: hmm, not sure we care about that case

[the remark this answers is not in the document]
Comment (trinity-1686a) [reply quickwit/quickwit-search/src/leaf.rs:1300]: would be nice to fix, but it's not actually changing ordering, so not a big deal imo

File: quickwit/quickwit-search/src/leaf.rs [hunk 5]
```diff
@@ -2380,302 +2401,6 @@ mod tests {
...
-        remove_redundant_timestamp_range(&mut request_direct, split, timestamp_field);
-        let expected_direct = expected
-            .clone()
-            .map(bool_filter)
-            .unwrap_or(QueryAst::MatchAll);
-        assert_ast_eq(&request_direct, &expected_direct);
-    }
-
-    #[test]
-    fn test_remove_timestamp_range() {
```
Comment (trinity-1686a) [inline quickwit/quickwit-search/src/leaf.rs:2430]: are there equivalents to the tests being removed? they were testing edge cases, and for some are even regression tests

Comment (PSeitz) [reply quickwit/quickwit-search/src/leaf.rs:2430]: restored the tests

**Review by trinity-1686a — APPROVED** [review]
